include/RAJA/util/basic_mempool.hpp

```diff
@@ -0,0 +1,381 @@
+
+      if ( found != m_used_space.end() ) {
+
+        add_free_chunk(found->first, found->second);
+
+        m_used_space.erase(found);
+
+      } else {
+        fprintf(stderr, "Invalid free %p", ptr);
+        std::abort();
+      }
+
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+private:
+  struct memory_chunk {
+    void* begin;
+    void* end;
+  };
+
+  void add_free_chunk(void* begin, void* end)
+  {
+    // integrates a chunk of memory into free_space
+    free_type::iterator invl = m_free_space.end();
+    free_type::iterator next = m_free_space.lower_bound(begin);
+
+    // check if prev exists
+    if (next != m_free_space.begin()) {
+      // check if prev can cover [begin, end)
+      free_type::iterator prev = next; --prev;
+      if (prev->second == begin) {
+        // extend prev to cover [begin, end)
+        prev->second = end;
+
+        // check if prev can cover next too
+        if (next != invl) {
+          assert(next->first != begin);
+
+          if (next->first == end) {
+            // extend prev to cover next too
+            prev->second = next->second;
+
+            // remove redundant next
+            m_free_space.erase(next);
+          }
+        }
+        return;
+      }
+    }
+
+    if (next != invl) {
+      assert(next->first != begin);
+      
+      if (next->first == end) {
+        // extend next to cover [begin, end)
+        m_free_space.insert(next, free_value_type{begin, next->second});
+        m_free_space.erase(next);
+
+        return;
+      }
+    }
+
+    // no free space adjacent to this chunk, add seperate free chunk [begin, end)
+    m_free_space.insert(next, free_value_type{begin, end});
+  }
+
+  void remove_free_chunk(free_type::iterator iter, void* begin, void* end) {
+
+    void* ptr = iter->first;
+    void* ptr_end = iter->second;
+
+    // fixup m_free_space, shrinking and adding chunks as needed
+    if (ptr != begin) {
+
+      // shrink end of current free region to [ptr, begin)
+      iter->second = begin;
+
+      if (end != ptr_end) {
+
+        // insert free region [end, ptr_end) after current free region
+        free_type::iterator next = iter; ++next;
+        m_free_space.insert(next, free_value_type{end, ptr_end});
+      }
+
+    } else if (end != ptr_end) {
+
+      // shrink beginning of current free region to [end, ptr_end)
+      free_type::iterator next = iter; ++next;
+      m_free_space.insert(next, free_value_type{end, ptr_end});
+      m_free_space.erase(iter);
+
+    } else {
+
+      // can not reuse current region, erase
+      m_free_space.erase(iter);
+    }
+  }
+
+  void add_used_chunk(void* begin, void* end)
+  {
+    // simply inserts a chunk of memory into used_space
+    m_used_space.insert(used_value_type{begin, end});
+  }
+
+  memory_chunk m_allocation;
+  free_type m_free_space;
+  used_type m_used_space;
+};
+
+} /* end namespace detail */
+
+
+
+template < typename allocator_t >
+class mempool {
```

Comment:
New classes need Doxygen, and I think we are naming things CamelCase (`MemPool`)